apps/wallet-mobile/src/yoroi-wallets/hooks/index.ts

```diff
@@ -985,34 +994,83 @@ export const useNativeAssetImage = ({
+  responseType = 'binary',
 }: NativeAssetImageRequest) => {
   const network = networkId === 300 ? 'preprod' : 'mainnet'
   const pWidth = PixelRatio.getPixelSizeForLayoutSize(Number(width))
   const pHeight = PixelRatio.getPixelSizeForLayoutSize(Number(height))
-  const isMediaTypeSupported = supportedTypes.includes(mediaType.toLocaleLowerCase())
+  const lcMediaType = mediaType.toLocaleLowerCase()
+  const isMediaTypeSupported = supportedTypes.includes(lcMediaType)
+  const needsGif = lcMediaType === 'image/gif' && Platform.OS === 'ios'
+  const mimeType = needsGif ? 'image/gif' : 'image/webp'
+  const headers = useMemo(
+    () => ({
+      Accept: responseType === 'binary' ? mimeType : 'text/plain',
+      'X-Encoded-Mimetype': mimeType,
+    }),
+    [mimeType, responseType],
+  )
+  const queryClient = useQueryClient()
+
+  const [isError, setError] = React.useState(false)
+  const [isLoading, setLoading] = React.useState(false)
+
+  const queryKey = ['native-asset-img', policy, name, responseType, `${pWidth}x${pHeight}`, contentFit]
+
   const query = useQuery({
+    enabled: isMediaTypeSupported,
     staleTime: Infinity,
-    queryKey: ['native-asset-img', policy, name, `${width}x${height}`, resizeMode],
-    queryFn: async () => {
-      const response = await fetch(
-        `https://${network}.cardano-nativeassets-prod.emurgornd.com/${policy}/${name}?width=${pWidth}&height=${pHeight}&kind=${kind}&fit=${resizeMode}`,
-      )
+    queryKey,
+    queryFn: async (context) => {
+      const count = queryClient.getQueryState(context.queryKey)?.dataUpdateCount
+      const cache = count ? `&cache=${count}` : ''
+      const requestUrl = `https://${network}.cardano-nativeassets-prod.emurgornd.com/${policy}/${name}?width=${pWidth}&height=${pHeight}&kind=${kind}&fit=${contentFit}${cache}`
+
+      if (responseType === 'binary') {
+        setLoading(true)
+        setError(false)
+        return requestUrl
+      }
+
+      const response = await fetch(requestUrl, {
+        headers,
+      })
       if (!response.ok) {
-        throw new Error(`NativeAsset CDN response was not ok for policy=${policy} name=${name}`)
+        throw new Error(`NativeAsset CDN: Not ok - ${requestUrl}`)
       }
       if (response.status === 201 || response.status === 202) {
-        throw new Error(`NativeAsset CDN still processing policy=${policy} name=${name}`)
+        throw new Error(`NativeAsset CDN: Processing - ${requestUrl}`)
       }
-      return `data:image/webp;base64,${await response.text()}`
+      return `data:${mimeType};base64,${await response.text()}`
     },
-    enabled: isMediaTypeSupported,
   })
 
+  const timerRef = React.useRef<any>(0)
+  React.useEffect(() => () => clearTimeout(timerRef.current), [])
+
+  const onError = useCallback(() => {
+    setError(true)
+    const count = queryClient.getQueryState(queryKey)?.dataUpdateCount
+    if (count && count < 10) {
+      timerRef.current = setTimeout(query.refetch, count * 300)
```

Comment:
Mmm... could be, I only want it refetching when there's an error though. And it is possible to have an image, no errors, and then invalidate it and get an error.
Not sure how useQuery  would react to refetchInterval being modified. Or if it calls it again after having returned false before...

Reply:
Also, `query.state.error` wouldn't work, because the query didn't fail. The query generates the url successfully. 
It's the <Image/> component the one doing the image fetching and possibly failing. That's why I have to use the onError callback to see if it failed.
On the base64 version, queryFn can fail and react-query will automatically retry a few times before actually showing the error. That's what I'm imitating for the binary version with the timeouts.